### PR TITLE
feat(security): client_negotiation handle mechanism selected response

### PR DESCRIPTION
**Pull request by levy5307**
client_negotiation handle mechanism selected response

### CODE REVIEW

File: include/dsn/utility/error_code.h [hunk 1]
```diff
@@ -124,4 +124,5 @@ DEFINE_ERR_CODE(ERR_UNAUTHENTICATED)
 DEFINE_ERR_CODE(ERR_KRB5_INTERNAL)
 
 DEFINE_ERR_CODE(ERR_SASL_INTERNAL)
+DEFINE_ERR_CODE(ERR_NOT_COMPLEMENTED)
```
Comment (neverchanje) [inline include/dsn/utility/error_code.h:127]: What ERR_NOT_COMPLEMENTED suggests? I didn't see the meaning anywhere.

File: src/runtime/security/sasl_client_wrapper.cpp [hunk 2]
```diff
@@ -26,16 +28,39 @@ DSN_DECLARE_string(service_name);
 
 error_s sasl_client_wrapper::init()
 {
-    // TBD(zlw)
-    return error_s::make(ERR_OK);
+    FAIL_POINT_INJECT_F("sasl_client_wrapper_init", [](dsn::string_view str) {
+        if (0 == str.compare("ERR_OK")) {
```
Comment (neverchanje) [inline src/runtime/security/sasl_client_wrapper.cpp:32]:
```suggestion
  return error_code::try_get(str, ERR_UNKNOWN);
```

File: include/dsn/utility/error_code.h (hunk 1 above)
Comment (levy5307) [reply include/dsn/utility/error_code.h:127]: `ERR_NOT_COMPLEMENTED` means the security process is not complemented, that we should process the next step by `sasl->step`.
First I used the `ERR_NOT_IMPLEMENTED`, but the meaning of this error code is not accure  accurate  for this case

Comment (acelyc111) [reply include/dsn/utility/error_code.h:127]: How about `ERR_SASL_INCOMPLETE` ?

File: src/runtime/security/client_negotiation.cpp [hunk 3]
```diff
@@ -111,6 +107,41 @@ void client_negotiation::on_recv_mechanisms(const negotiation_response &resp)
...
+{
+    if (!check_status(resp.status, negotiation_status::type::SASL_SELECT_MECHANISMS_RESP)) {
+        fail_negotiation();
+        return;
+    }
+
+    // init client sasl
+    auto err_s = _sasl->init();
+    if (!err_s.is_ok()) {
+        dwarn_f("{}: initiaze sasl client failed, error = {}, reason = {}",
```
Comment (acelyc111) [inline src/runtime/security/client_negotiation.cpp:120]:
```suggestion
        dwarn_f("{}: initialize sasl client failed, error = {}, reason = {}",
```

File: src/runtime/security/server_negotiation.cpp [hunk 4]
```diff
@@ -60,52 +60,44 @@ void server_negotiation::handle_request(negotiation_rpc rpc)
...
-                enum_to_string(request.status),
-                negotiation_status::type::SASL_SELECT_MECHANISMS);
+                err_s.code().to_string(),
+                err_s.description());
         fail_negotiation();
         return;
     }
+
+    negotiation_response &response = rpc.response();
+    _status = response.status = negotiation_status::type::SASL_SELECT_MECHANISMS_RESP;
```
Comment (acelyc111) [inline src/runtime/security/server_negotiation.cpp:100]: Good refactor!

File: include/dsn/utility/error_code.h (hunk 1 above)
Comment (levy5307) [reply include/dsn/utility/error_code.h:127]: good idea.

**Review by neverchanje — APPROVED** [review]

**Review by acelyc111 — APPROVED** [review]

**Pull request merged**